core/commands/pin.go

```diff
@@ -282,3 +272,76 @@ type RefKeyObject struct {
 type RefKeyList struct {
 	Keys map[string]RefKeyObject
 }
+
+func pinLsKeys(args []string, typeStr string, ctx context.Context, n *core.IpfsNode) (map[string]RefKeyObject, error) {
+
+	keys := make(map[string]RefKeyObject)
+
+	for _, p := range args {
+		dagNode, err := core.Resolve(ctx, n, path.Path(p))
+		if err != nil {
+			return nil, err
+		}
+
+		k, err := dagNode.Key()
+		if err != nil {
+			return nil, err
+		}
+
+		pinType, pinned, err := n.Pinning.IsPinnedWithType(k, typeStr)
+		if err != nil {
+			return nil, err
+		}
+
+		if !pinned {
+			return nil, fmt.Errorf("Path '%s' is not pinned", p)
+		}
+
+		switch pinType {
+		case "direct", "indirect", "recursive", "internal":
+		default:
+			pinType = "indirect through " + pinType
```

Comment:
@rht if you ask why the `switch pinType { ... }` is not part of `IsPinnedWithType` then my answer is that it would be inefficient for `IsPinnedWithType` users which are internal functions to have to parse "indirect through XXX" if they want to get XXX.

If there are many `IsPinnedWithType` users that want "indirect through XXX" then we can add `IsPinnedWithTypePretty` that just calls `IsPinnedWithType` and does what the `switch pinType { ... }` does.